subprojects/base-services/src/main/java/org/gradle/internal/classloader/FilteringClassLoader.java

```diff
@@ -65,13 +69,13 @@
 
     public FilteringClassLoader(ClassLoader parent, Spec spec) {
         super(parent);
-        packageNames.addAll(spec.packageNames);
-        packagePrefixes.addAll(spec.packagePrefixes);
-        resourceNames.addAll(spec.resourceNames);
-        resourcePrefixes.addAll(spec.resourcePrefixes);
-        classNames.addAll(spec.classNames);
-        disallowedClassNames.addAll(spec.disallowedClassNames);
-        disallowedPackagePrefixes.addAll(spec.disallowedPackagePrefixes);
+        packageNames = new HashSet<String>(spec.packageNames);
+        packagePrefixes = new TrieSet(spec.packagePrefixes);
```

Comment:
I was expecting that you would be using `FileHierarchySet` (which would need to be adjusted to support `.` as separator). `Trie` is a character based trie, so we have a node per char. I think a node per segment would be better.

If it works like it is now and is faster, we can also use this implementation.

Reply:
`FileHierarchySet` works on absolute paths, so it would take quite a bit of work to make that work for these abstract paths. I think `Trie` should be pretty fast in itself. Indeed, it should be faster than breaking things up into segments and then comparing each segment as we go along.